[Elumalai Oviya Dharshini] iP

ovidharshini opened the pull request:
# Handbook
Are you behind on catching up with your last few commits these days? Does having to connect to the internet to log onto your favourite cloud-based task manager bother you? Are you tired of manually formatting your tasks on a new note on your smartphone
- like 
- this
- every other day?

Or is the thought of slow...ly dragging your mouse across the screen a few dozen times to update your weekly list of things to be done putting you off updating your weekly list of things to be done?

### If so, Handbook is the right app for you!
Handbook is a task manager that is free, lightweight, and easy to use. 
It is a heavily text-reliant application that seeks to meet the needs of the general advanced user. 

## What Handbook does
- [x] Manage tasks and deadlines 
- [x] Save tasks onto your computer in an **easily accessible format** 
- [ ] Reminders (coming soon)
- [ ] Syncing with common note-taking applications (on the todo list!)
- [ ] ~~Do your work for you~~

## Why Handbook
Here is what some of our users have to say about Duke
> _"Handbook is a fantastic application - I wish I could have used Handbook to manage my schedule when I was coding Handbook... Give it a go, you won't regret it." - App Developer_

> _"Why this over the other 270 pull requests, you ask? Good question. Check out the other applications too. But before that, try out Handbook." - Anonymous User_

## Installing Handbook
So, you've decided that you want to see what the hype is all about. What now?
All you need to do is
1. download the jar file from [here](https://github.com/ovidharshini/ip/releases/tag/A-Jar)
2. move it to where you want the application to be accessed from
2. double click it
3. and let it manage your tasks for you :star2:

If you are a Java programmer, you can customize Handbook to your liking as well; give it your own personality. Here's a brief snippet of the `Ui` class:
```java
public class Ui {
    Scanner reader = new Scanner(System.in);
    /**
     * Displays a goodbye message and closes the associated Scanner.
     */
    public void showGoodbye() {
        System.out.println("Bye, hope to see you again soon!");
        reader.close();
    }
    ...
}
```

Commit:
Add Gradle support

ovidharshini committed:
added user echo

ovidharshini committed:
Store input list

ovidharshini committed:
Represent task as class

ovidharshini committed:
Add marking for tasks

ovidharshini committed:
Add todo, deadline, event tasks

ovidharshini committed:
Add files for I/O redirection testing

ovidharshini committed:
Add input sanitization

ovidharshini committed:
Add delete feature for list

ovidharshini committed:
Add save task feature

ovidharshini committed:
Add date, time recognition

ovidharshini committed:
Merge branch 'branch-Level-7'

ovidharshini committed:
Merge branch 'branch-Level-8'

ovidharshini committed:
Refactor to use more OOP-based code

ovidharshini committed:
Add packages

ovidharshini committed:
Add JUnit tests

ovidharshini committed:
Apply Java coding standards

ovidharshini committed:
Add keyword search command

ovidharshini committed:
Merge branch 'branch-Level-9'

ovidharshini committed:
Resolve merge conflict in import packages

ovidharshini committed:
Resolve merge conflict

ovidharshini committed:
Add JavaDoc comments

ovidharshini committed:
Resolve merge conflict

ovidharshini committed:
Add comments for search functionality

tobihy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,124 @@
+        boolean isFirst = true;
+        for (int i = 0; i < tasks.size(); i++) {
+            String s = isFirst ? "" : "\n";
+            f.write(s + tasks.get(i).writeToFile());
```
Perhaps a line break after the "." operator would enhance the readability?

tobihy replied:
Perhaps a line break after the "." operators would enhance readability? I noticed the same issue in several other places too.

tobihy commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+     * @return formatted string of all Tasks in TaskList and number of Tasks in TaskList delimited by newlines
+     */
+    public String toString() {
+        String s = "";
+        for (int i = 0; i < tasks.size(); i++) {
+            if (!tasks.get(i).isEmpty()) {
+                int index = i + 1;
+                s += "  " + index + ". " + tasks.get(i) + "\n";
+            }
+            else {
+                s += "You have " + i +" tasks on your list.";
+                break;
+            }
+        }
+        if (tasks.isEmpty()) {
+            s =  "You now have 0 tasks on your list.";
+        }
+
+        return s;
```
Maybe try adding line breaks in between logical units present in the code? 

In addition, if-else statements should have the following form:
```
if (condition) {
    statements;
} else {
    statements;
}
```

ovidharshini replied:
Thanks for highlighting this! I have now changed the code to follow the Java coding standards more closely

tobihy commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,22 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
Minor nitpick, but import statements here should be reordered as per the [Java coding standard](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements). I noticed similar issues in other places too.

ovidharshini replied:
I had a look at this and the guideline mentions that "the ordering of import statements must be consistent" but is otherwise generally vague. The consensus on a few StackOverflow posts appears to be that they ought to be in order of importance, then alphabetical order, with custom packages being imported after other packages. Is there another - perhaps more definitive - source you can point me to with regards to this?

tobihy commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,22 @@
+public class ParserTest {
+    @Test
+    public void testParsing() {
+        assertEquals("", Parser.parse("hi"));
+        assertEquals("", Parser.parse("todo"));
+        assertEquals("", Parser.parse("todo "));
+        assertEquals("", Parser.parse("event "));
+        assertEquals("", Parser.parse("event birthday at"));
+        assertEquals("", Parser.parse("event birthday at "));
+        assertEquals("event", Parser.parse("event birthday at 1500"));
+        assertEquals("", Parser.parse("deadline cs2103t week 3 ip tasks by"));
+        assertEquals("", Parser.parse("deadline cs2103t week 3 ip tasks by "));
+        assertEquals("deadline", Parser.parse("deadline cs2103t week 3 ip tasks by 2022/01/27 2359"));
```
Consider splitting the test into multiple JUnit test methods?

tobihy reviewed (commented):
I see that you like Genshin Impact a lot 💯 

Overall, I found your code easy to read and understand, except for a few places where the formatting could be improved. I noticed a few minor coding standard violations too. Other than that, LGTM. Keep up the good work! 👍

chuashihong commented on src/main/java/duke/task/TaskWithDateTime.java:
```diff
@@ -0,0 +1,90 @@
+ * @author  Elumalai Oviya Dharshini
+ * @version 0.1
+ */
+public class TaskWithDateTime extends Task {
```
I like how you make another TaskWithDateTime class that differentiate between normal task and task with date and time!

chuashihong commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class DeadlineTest {
+    @Test
+    public void testStringConversion() {
```
Clear method name, it is obvious what is going to tested in this method

chuashihong commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    @Test
+    public void testWriteToFile() {
```
Clear method name, it is obvious what is going to tested in this method!

chuashihong commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,39 @@
+     * @return formatted string of Todo, its completion status and a Todo marker
+     */
+    @Override
+    public String writeToFile() {
```
One suggestion: Maybe you can rename writeToFile as taskToStringFormat ? Because write to file at the first sight, it is going to do the whole thing writing to file, but it is just producing String format of the task to be written to file.
But it is also good that you use writeToFile (no harm)

chuashihong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,124 @@
+                String[] tmp = line.split("\\|");
+                boolean isDone = tmp[1].trim().equals("D");
+
+                switch (tmp[0].trim()) {
```
You followed the coding standard for switch case! Great Job!

chuashihong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,124 @@
+    //Reused from https://stackoverflow.com/questions/27904329/warning-file-mkdir-is-ignored#answer-54341862
+    // with minor modifications
+    @SuppressWarnings("unused")
+    private static void IGNORE_RESULT(boolean result) {
```
I am not sure what does this means, it is not returning anything

ovidharshini replied:
This utility function is defined for ```mkdirs``` - which creates parent directories of a given location (if they don't exist) and returns ```True``` or ```False``` depending on whether directories were created. This suppresses the ```File.mkdirs() is ignored``` warning that results from not handling the return value of ```mkdirs```. 

I am not certain if this is the recommended way to handle it so I am leaving it as is for now.

chuashihong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,111 @@
+                System.out.println("Oops, you need to format deadline tasks as \"deadline X by Y\" :c");
+                return "";
+            }
+            String lastWord = input.substring(input.lastIndexOf(" ") + 1);
+            if (lastWord.equals("by")) {
+                return "";
+            }
+            return command;
+        }
+
+        // Handle event
+        if (firstWord.equals("event")) {
+            input = input.substring(5).trim();
+            if (!input.contains(" at ")) {
+                System.out.println("Oops, you need to format event tasks as \"event X at Y\" :c");
+                return "";
+            }
+            String lastWord = input.substring(input.lastIndexOf(" ")+1);
+            if (lastWord.equals("at")) {
+                return "";
+            }
+            return command;
+        }
+
+        // Handle find
+        if (firstWord.equals("find")) {
+            input = input.substring(4).trim();
+            if (input.equals("")) {
+                System.out.println("Oops, you need to mention what the keyword is :c");
+                return "";
+            }
+            if (input.contains(" ")) {
+                System.out.println("Oops, you can only search for one keyword at a time :c");
+                return "";
+            }
+            return command;
+        }
+
+        return "";
+    }
```
How about making these logic sentence into a switch case statement? So that it is easier to follow and read!

ovidharshini replied:
Thanks for highlighting this; I have now changed the code to make use of switch-case statements to make it more readable.

chuashihong reviewed (commented):
The whole code looks very clear and I know what is every method doing, the structure of the project is also neat and clean! Keep up the good work!

ovidharshini committed:
Resolve coding standard violations

ovidharshini committed:
Merge remote-tracking branch 'origin/add-gradle-support' into HEAD

ovidharshini committed:
Add gradle support

ovidharshini committed:
Resolve checkstyle warnings

ovidharshini committed:
Delete redundant code

ovidharshini committed:
Add GUI

ovidharshini committed:
Merge branch 'branch-Level-10'

ovidharshini committed:
Use Java assert feature

There are important assumptions that should hold at various points
in the code.

Using Java assertions documents these assumptions and improves code
quality.

ovidharshini committed:
Improve code quality

ovidharshini committed:
Merge pull request #3 from ovidharshini/branch-A-CodeQuality

Improve code quality

ovidharshini committed:
Merge branch 'master' into branch-A-Assertions

ovidharshini committed:
Merge pull request #2 from ovidharshini/branch-A-Assertions

Use assertions

ovidharshini committed:
Fix bug with assertions

ovidharshini committed:
Add DoAfter tasks

ovidharshini committed:
Handle NumberFormatException in input

ovidharshini committed:
Handle assertion edgecase

ovidharshini committed:
Merge branch 'branch-B-DoAfterTasks'

ovidharshini committed:
Decrease font size

ovidharshini committed:
Fix minor spelling error

ovidharshini committed:
Add tick unicode symbol

ovidharshini committed:
Add icon

ovidharshini committed:
Add representative screenshot

ovidharshini committed:
Fix date storage bug

ovidharshini committed:
Add user guide

ovidharshini committed:
Set theme jekyll-theme-minimal

ovidharshini committed:
Set theme jekyll-theme-slate

ovidharshini committed:
Set theme jekyll-theme-slate

ovidharshini committed:
Set theme jekyll-theme-midnight

ovidharshini committed:
Set theme jekyll-theme-slate

ovidharshini committed:
Set theme jekyll-theme-time-machine

ovidharshini committed:
Set theme jekyll-theme-cayman

ovidharshini committed:
Set theme jekyll-theme-tactile

ovidharshini committed:
Set theme jekyll-theme-hacker

ovidharshini committed:
Delete theme

ovidharshini committed:
Set markdown processor for gh pages

ovidharshini committed:
Update _config.yml

ovidharshini committed:
Fix bug with GFMD rendering

ovidharshini committed:
Change Ui.png capitalization

ovidharshini committed:
Add Ui.png

ovidharshini committed:
Add Mac, Linux compatibility

ovidharshini committed:
Change minor messages

ovidharshini committed:
Resolve filepath bug, GUI errors

ovidharshini committed:
Add updated Ui.png

ovidharshini committed:
Add attribution for images

ovidharshini committed:
Resolve bug with image display
